### ISSUE TITLE
Question regarding storage on multi-node install

### ISSUE THREAD

**Issue by dajester2013** [ubuntu/microk8s#1597]
I set up a multi-node install of microk8s 1.19, and enabled storage.

However, when I try to deploy a statefulset with a volume template, I run into issues.

The server pod can be brought up on a different node from the one the PV is actually created on, causing the pod to fail to start, as the directory for the volume does not exist.  Is this expected?  How can I ensure the volume is provisioned on the correct node?

Here's the relevant bit from my helm template with the volume claim:
``` yaml
apiVersion: apps/v1
kind: StatefulSet
metadata:
  name: {{ $namePrefix }}-postgres
  labels:
    {{ include "baseline-study.selectorLabels" . | nindent 4 }}
spec:
  replicas: 1
  serviceName: {{ $namePrefix }}-studydb
  updateStrategy:
    type: RollingUpdate
  selector:
    matchLabels:
      app: postgres
      tier: database
      {{- include "baseline-study.selectorLabels" . | nindent 6 }}
    
  template:
    metadata:
      labels:
        app: postgres
        tier: database
        {{- include "baseline-study.selectorLabels" . | nindent 8 }}
    spec:
      {{- if .Values.postgres.configFile }}
      {{ $pgconf := .Values.postgres.configFile }}
      {{- if (len (.Files.Get $pgconf)) }}
      volumes:
        - name: {{ .Release.Name }}-pgconfig
          configMap:
              name: {{ .Release.Name }}-pgconfig
      {{ end -}}
      {{ end -}}
      terminationGracePeriodSeconds: 10

      containers:
        - name: postgres
          image: postgres:12.4
          
          {{- if .Values.postgres.configFile }}
          {{ $pgconf := .Values.postgres.configFile }}
          {{- if (len (.Files.Get $pgconf)) }}
          args: 
           - -c
           - config_file=/etc/postgres.conf
          {{ end -}}
          {{ end -}}
          
          imagePullPolicy: IfNotPresent

          ports:
            - name: postgres
              containerPort: 5432
              protocol: TCP

          resources:
            requests:
              cpu: 100m     # TODO: is this right???
              memory: 512Mi # TODO: is this right???

          env:
            - name: POSTGRES_USER
              valueFrom:
                secretKeyRef:
                  key: pg.username
                  name: {{ $pgCredSecretName }}
            - name: PGUSER
              valueFrom:
                secretKeyRef:
                  key: pg.username
                  name: {{ $pgCredSecretName }}
            - name: POSTGRES_PASSWORD
              valueFrom:
                secretKeyRef:
                  key: pg.password
                  name: {{ $pgCredSecretName }}
            - name: POSTGRES_DB
              value: studydb
            - name: PGDATA
              value: /var/lib/data/postgresql/data/pgdata
            - name: POD_IP
              valueFrom:
                fieldRef:
                  apiVersion: v1
                  fieldPath: status.podIP


          livenessProbe:
            exec:
              command: ["sh","-c","exec pg_isready --host $POD_IP"]
            failureThreshold: 6
            initialDelaySeconds: 60
            periodSeconds: 10
            successThreshold: 1
            timeoutSeconds: 5

          readinessProbe:
            exec:
              command: ["sh","-c","exec pg_isready --host $POD_IP"]
            failureThreshold: 3
            initialDelaySeconds: 5
            periodSeconds: 5
            successThreshold: 1
            timeoutSeconds: 3


          volumeMounts:
            - mountPath: /var/lib/postgresql/data/pgdata
              name: {{ $namePrefix }}-data-vol
              subPath: postgres-db
            {{- if .Values.postgres.configFile }}
            {{ $pgconf := .Values.postgres.configFile }}
            {{- if (len (.Files.Get $pgconf)) }}
            - name: {{ .Release.Name }}-pgconfig
              mountPath: /etc/postgres.conf
              subPath: postgres.conf
            {{ end -}}
            {{ end -}}

            {{ $files := .Files.initdb }}
            {{ range $files }}
            - name: {{ $namePrefix }}-initdb
              mountPath: /docker-entrypoint-initdb.d/{{ . }}
              subPath: {{ . }}
            {{ end }}
  volumeClaimTemplates:
    - metadata:
        name: {{ $namePrefix }}-data-vol
        labels: {{ include "baseline-study.selectorLabels" . | nindent 10 }}
      spec:
        accessModes:
          - ReadWriteOnce
        resources:
          requests:
            storage: 10G
```

**Comment by balchua** [comment]
@dajester2013 the storage addon is a hostpath based storage.  Not really for multi node cluster.

There are other storage solutions that is more suited for multi nodes.  Such as rook (ceph), openebs, portworx, etc...

I have tried rook with ceph and i find it difficult to work with.  Its also very much possible that im just not familiar with these.

@stubbss did try openebs. He may have some feedback about it.

**Comment by sstubbs** [comment]
Hi @dajester2013 yes there are quite a few options. Openebs has worked well for me. Their local pv is probably what you looking for https://docs.openebs.io/docs/next/localpv.html

**Comment by dajester2013** [comment]
> Hi @dajester2013 yes there are quite a few options. Openebs has worked well for me. Their local pv is probably what you looking for https://docs.openebs.io/docs/next/localpv.html

Thanks - I'll give that a shot.  I'm still a bit new to the kubernetes world.

**Comment by Namyts** [comment]
@sstubbs Have you managed to get OpenEBS cStor working on microk8s? I got the localPV to work relatively easily, but am encountering [an error](https://stackoverflow.com/questions/64170522) with cStor. I was wondering if it was an issue with microk8s, OpenEBS, or a fault of my own...

**Comment by sstubbs** [comment]
Hi @Namyts. I did use cstore as well as ceph via rook for a few months. Both of them fairly straightforward to setup from what I remember but did need a bit of altering of configuration. Now I only use openebs localpv. 

For anything that needs clustered storage I now use minio operator with their direct csi driver. Not because of issues with openebs or rook ceph but because I like a lot of things that minio does. They busy reworking the csi driver to handle different storage classes. Once that's done I will probably use it for everything not just minio if it works well https://github.com/minio/direct-csi as I'm trying not to have too many different projects doing the same thing.

I never tested it with loop devices instead of raw block devices. That could potentially be a problem. openebs have a great slack channel. Perhaps ask there.

**Comment by dajester2013** [comment]
So, just installed openebs using helm3, but any attempt to create a PVC results in a timeout error:

StorageClass manifest (copied from openebs local pv docs):
```
apiVersion: storage.k8s.io/v1
kind: StorageClass
metadata:
  name: local-hostpath
  annotations:
    openebs.io/cas-type: local
    cas.openebs.io/config: |
      - name: StorageType
        value: hostpath
      - name: BasePath
        value: /var/local-hostpath
provisioner: openebs.io/local
reclaimPolicy: Delete
volumeBindingMode: WaitForFirstConsumer
```

PVC manifest:
```
kind: PersistentVolumeClaim
apiVersion: v1
metadata:
  name: local-hostpath-pvc
spec:
  storageClassName: local-hostpath
  accessModes:
    - ReadWriteOnce
  resources:
    requests:
      storage: 5G
```

This is the error I get when trying to apply the PVC manifest:
```
Internal error occurred: failed calling webhook "admission-webhook.openebs.io": Post "https://admission-server-svc.openebs.svc:443/validate?timeout=5s": context deadline exceeded
```

**Comment by panlinux** [comment]
> @dajester2013 the storage addon is a hostpath based storage. Not really for multi node cluster.

Can't this be alleviated a bit by setting `volumeBindingMode` to `WaitForFirstConsumer`? The docs say that if `volumeBindingMode` is not specified, then `Immediate` is used, and that can lead to the PV being created in one node and the pod in another one:

>    volumeBindingMode	<string>
>      VolumeBindingMode indicates how PersistentVolumeClaims should be
>      provisioned and bound. When unset, VolumeBindingImmediate is used. This
>      field is only honored by servers that enable the VolumeScheduling feature.

In https://github.com/ubuntu/microk8s/blob/83f24a689c57b1c729c72cf0d4d6df82906e8be6/microk8s-resources/actions/storage.yaml#L40 it's not set, meaning the default `Immediate` value is used then.

**Comment by panlinux** [comment]
For what is worth, rancher's `local-path` provisioner sets it to the correct value:
```
ubuntu@f1-k3s:~$ kubectl get sc local-path -o yaml | grep ^volumeBindingMode:
volumeBindingMode: WaitForFirstConsumer
```
It's still kind of a workaround, since obviously the volume is still local to the node.

**Comment by balchua** [comment]
We have included OpenEBS addon from MicroK8s 1.21, hence closing the issue.
Thanks!
